content/blogposts/2022/multi-version-upgrades.md

```diff
@@ -0,0 +1,109 @@
+
+In what is likely the largest departure from golang-migrate, we allowed migrations to [define multiple parents](https://github.com/sourcegraph/sourcegraph/pull/30664). This allows our migration definitions to form a [directed acyclic graph](https://en.wikipedia.org/wiki/Directed_acyclic_graph) (instead of an absolutely ordered list). Note that in the image above our migrations no longer form a gapless sequence, and a [topological sort](https://en.wikipedia.org/wiki/Topological_sorting) over the migration graph is done to determine an application order.
+
+Well that seemed like a bunch of work for no payoff. **Why do that at all?** Short answer: engineering ergonomics. When we had a gapless sequence we would _frequently_ run into merge conflicts when two engineers added migrations around the same time. Since each branch did not contain the other's migration, they would both choose the same sequence number. Because of limitations in our CI (no build queues and no non-stale base branch merge requirement), this conflict would, the majority of the time, be caught only after both branches had been merged and validation errors turned the `main` branch to go red.
+
+The following image shows a reduced (but fairly accurate) migration graph over several minor releases. When a new migration is created, it chooses its own parents: the set of migration definitions that do not yet have any children. In this formulation, if two engineers add migrations around the same time, they may share a parent, but they also have distinct identifiers (with very high probability). They don't conflict, and should be able to be run in either order relative to one another as they were both developed in independent branches. Setting parents in this manner allows us to track explicit dependencies: the set of migrations that were already defined _at the time the new migration was created_.
+
+<Figure src="https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/mvu-graph-structures.png" alt="How migration graph structures change release after release." />
+
+## The pile of wrenches
+
+The evolution of our migration definitions over time has created a number of issues to overcome for multi-version upgrades. The biggest issue comes from periodically squashing migration definitions, which erases information necessary to the upgrade process. In the image above, there's no valid way to upgrade from 3.36 to 3.39 in one step. The squashed migration contains _both_ the SQL that's already been run by the 3.36 installation, as well as the SQL defined in 3.37 that has yet to be applied. As we cannot partially apply a single migration, we cannot support this upgrade path.
+
+To get around this, we need to undo the squash operation over time to get a single, consistent "stitched" migration graph.
+
+<Figure src="temp/timeline.png" alt="timeline" />
```

Comment:
@efritz What image should this be?

Reply:
https://github.com/sourcegraph/about/blob/8eda68889b3f6db9cb4638003ee6ca64e4a7c7b5/public/blog/temp/timeline.png

Reply:
```suggestion
<Figure src="https://storage.googleapis.com/sourcegraph-assets/blog/multi-version-upgrades/timeline.png" alt="timeline" />
```